### ISSUE TITLE
Isolation of custom styles

### ISSUE THREAD

**Issue by pieplu** [holidaypirates/nucleus#34]
Thanks for your project, it looks really good!

I have a request / bug. I think it will be good to isolate the custom style, to don't modify default nucleus styles.

Tested with **nucleus-styleguide@1.0.8**

For exemple, I have this code on my stylesheet:

```css
li {
  font-size: 1rem;
  line-height: 1.71429rem;
}
```
And it's broke the nucleus menu (purple bar and icon size):

![visualbug](http://img11.hostingpics.net/pics/583912nucleusalineheight.png "visualbug")

An idea is to use iframe, and have our custom styles only on iframe (like [styleguide.js](https://rawgit.com/EightMedia/styleguide.js/master/test/expected/index.html)  or patternlab).

**Comment by Rhym** [comment]
I can see the benefits of having the styles isolated. I'm flying to NZ for two weeks, but I could work on this when I get back. @michaelseibt if you agree that it would be the right direction?

**Comment by pkkummermo** [comment]
One might argue that allowing styling of elements is a bad thing and should be avoided. I highly recommend looking into BEM as methodology when creating styles as you'll avoid things like this :)

**Comment by Rhym** [comment]
Well this repo is literally for the atomic methodology, which is inline with BEM. It still doesn't account for some global reset stuff that we carry across to most projects. There are a couple of generated styles that are effected too, for example in the code prettifier there is a class `.tag`, so you better not have an atom with the same name, or it throws a fit 😄

**Comment by pieplu** [comment]
@pkkummermo I totaly  agree whith that, but sometime we can't (or it's a really harder way). 

You can't allways put class on the markup you have to style:
- CMS and "user generated HTML" by rich text editor 
- markup sended by X plugins/back-end-stuff/modules

For each case we can find a workaround to don't style directly a HTML element (parent class, hook on templates, etc...), but depends of your website/techno complexity, it can be better (time, maintenance, ...) to define style directly on a HTML element. 

I know it's a bad practice, but we can't always follow good practices (for many raison).

**Comment by pieplu** [comment]
An other problem with a no-isolation pattern:

If you have a "default" value for your typography ( on html for exemple) with your base font-family, base font-size, etc...
This styles of nucleus will overwite it...
```css
.SG-Layout{
    font-family: Lato, sans-serif;
    font-size: 14px;
    ...
}
```

I don't think re-define a **font-family** and a **font-size** property on each modules, without use CSS inheritance is a good practice ...

**Comment by tomasz-nowakowski-stx** [comment]
Good idea, but instead of `iframe` please consider using _Shadow DOM_ and _custom elements_.

**Comment by adamziel** [comment]
@pieplu Those `font-family` and `font-size` declarations were really messing with my styleguide so I created this little script to isolate my components:

```js
(function() {
    const appResources = (
        [].concat(Array.from(document.querySelectorAll('head link[rel="stylesheet"]:not([href="styles/app.css"])')))
          .concat(Array.from(document.querySelectorAll('script link:not([href*="fixNucleus.js"]):not([href="styles/app.js"])')))
    );

    const previewElements = Array.from(document.querySelectorAll('.SG-component__preview'));
    previewElements.forEach(elem => {
        const children = elem.childNodes;
        const shadow = elem.attachShadow({mode: 'open'});
        appResources.forEach(r => shadow.appendChild(r.cloneNode()));
        children.forEach(c => shadow.appendChild(c));
    });

    appResources.forEach(r => r.parentNode.removeChild(r));
})();
```

Just save it as `js/fixNucleus.js` and add it to your config!

**Comment by adamziel** [comment]
Actually custom fonts does not work inside shadow dom in Chrome, so I used an alternative solution:
```js
(function() {
    var sheet = document.styleSheets[0];
    var rules = sheet.cssRules || sheet.rules;
    Array.from(rules).forEach(rule => {
        if(rule.selectorText === ".SG-layout") {
            rule.style.fontSize = '';
            rule.style.fontFamily = '';
        }
    });
})();
```

**Comment by macandcheese** [comment]
Are there plans to support this in a better way? I don't want, for instance, the .SG-layout default styles to overwrite by custom H4 class. I understand in the demo it has a namespace prefix and is using a div vs. the h4 element itself. Surely in the interest of semantics it would be better to be able to use h1 - h6 as they would appear in markup vs. div.SG-h1.

Seems like the underlying app.css used by Nucleus should at least be exposed to change, otherwise .SG-layout etc will just overwrite the custom styles, ha...
